### PR TITLE
Add Iterator and shortest/longest

### PR DESCRIPTION
**Pull request by rschmukler**

### CODE REVIEW

File: lib/enum/extra.ex [hunk 1]
```diff
@@ -322,4 +324,140 @@ defmodule Enum.Extra do
...
+      {_quick_enum, quick_count} when quick_count <  slow_count ->
+        longest_iter.enum
+
+      {quick_enum, quick_count}  when quick_count >= slow_count ->
+        case Enum.count(longest_iter) + slow_count do
+          x when x < quick_count  ->
+            quick_enum
+          x when x > quick_count  ->
+            longest_iter.enum
+          x when x == quick_count ->
```
Comment (glittershark) [inline lib/enum/extra.ex]: imo this would be a lot cleaner as a `cond`

File: lib/enum/extra.ex [hunk 2]
```diff
@@ -322,4 +324,140 @@ defmodule Enum.Extra do
...
+            Enum.find(list_of_enums, &(&1 in [quick_enum, longest_iter.enum]))
```
Comment (glittershark) [inline lib/enum/extra.ex]: Code style says not to use parens here

File: lib/enum/extra.ex [hunk 3]
```diff
@@ -322,4 +324,140 @@ defmodule Enum.Extra do
...
+
+    {longest_iter, slow_count} =
+      slow_counts
+      |> :lists.reverse()
+      |> Enum.map(&Iterator.new/1)
+      |> do_longest(0)
+
+    case longest_quick_count do
+      {_quick_enum, quick_count} when quick_count <  slow_count ->
+        longest_iter.enum
```
Comment (glittershark) [inline lib/enum/extra.ex]: do we need to call `enum` here? Isn't an `Iterator` an `Enum.t` anyway?

**Review by glittershark — CHANGES_REQUESTED** [review]
some rough stylistic stuff, looks pretty :bomb: overall

File: lib/enum/extra.ex (hunk 3 above)
Comment (rschmukler) [reply lib/enum/extra.ex]: Yes, but returning the same value as the input is the point of this function. ie. What if the input implements other protocols? It'd be annoying to lose that, nor have any information as to which input enumerable won out.

Comment (glittershark) [reply lib/enum/extra.ex]: oh, sure, that seems reasonable

**Review by glittershark — APPROVED** [review]

**Pull request merged**